Update GxGSV NMEA 4.10 - Add Signal averaging - Update gpsinformationwidget

bettellam opened the pull request:
**parse.c**
Update GxGSV - NMEA 4.10 `signal_id`

**sentence.h**
Add `signal_id`; //!< NMEA v4.1 - ID of the ranging signal

**qgsnmeaconnection.cpp**
NMEA RMC - Set qualitydescription only for status: A or V in  processRmcSentence
if the value in `status` is not `A or V` (e.g. D), for `qualitydescription` it is set with GGA

NMEA GSV - Add Signal averaging processGsvSentence
`QgsSatelliteInfo signal` is calculated as the average of the signals present in GxGSV
With future updates using `signal_id` you may get a result similar to this:
![Screenshot 2024-08-05 123331](https://github.com/user-attachments/assets/693e5208-0010-4877-899d-4c0e9c523035)

**qgsgpsinformationwidget.cpp**
Small updates in SkyPlot satellites (added color and different symbol for various constellations)

bettellam committed:
Update GxGSV - NMEA 4.10 signal_id

bettellam committed:
Add signal_id; //!< NMEA v4.1 - ID of the ranging signal

bettellam committed:
Update NMEA RMC -  GSV

NMEA RMC - Set qualitydescription only for status: A or V in  processRmcSentence
NMEA GSV - Add Signal averaging processGsvSentence

bettellam committed:
Update Skiplot satellites (QWTPOLAR)

bettellam committed:
Update Skiplot satellites (QWTPOLAR) - identation update

bettellam committed:
Update Skiplot satellites (QWTPOLAR) - identation update

bettellam committed:
Update Skiplot satellites (QWTPOLAR) - identation update

bettellam commented:
![Screenshot 2024-08-11 170749](https://github.com/user-attachments/assets/016703ea-1984-456e-b3c5-77634b1435c6)


![Screenshot 2024-08-12 110904](https://github.com/user-attachments/assets/59e6c97f-6792-452b-ad60-6a2617fc27fe)


![Screenshot 2024-08-11 171202 - Copia](https://github.com/user-attachments/assets/b599aa00-98ab-49ff-9c6d-fbffdf28a39b)

bettellam committed:
Update Skiplot satellites (QWTPOLAR) - update

bettellam committed:
Update Skiplot satellites (QWTPOLAR) - update

bettellam committed:
Update Skiplot satellites (QWTPOLAR) - update

bettellam committed:
Merge branch 'qgis:master' into patch-1

bettellam committed:
Update qgsgpsinformationwidget.cpp change marker size

bettellam commented:
@nyalldawson  Ciao
Sorry this request was not merged.

The change to **GxGSV** is a simple adaptation to NMEA 4.10

The change to **GxRMC** solves a problem with the new Unicore UM98x receivers, low-cost GNSS receivers that are having a considerable success on par with the UBlox F9P.
In particular with UM98x an improper use of the Status flag is performed in the NMEA GxRMC sentence whose standard values ​​are: **A=active or V=void**.
UM98x repeats the values ​​of the mode flag.

The current procedure in QGIS only checks Status=A to fix from the Mode flag:
`mLastGPSInformation.quality`
`mLastGPSInformation.qualityIndicator`

passing to Void any other value of Status and then:
`mLastGPSInformation.quality = Invalid` ;
`mLastGPSInformation.qualityIndicator = Invalid` ;
this makes the UM98x receivers unusable when they switch to RTK Fix mode.

It should be noted that
`mLastGPSInformation.quality `and
`mLastGPSInformation.qualityIndicator`
are set using NMEA **GxGGA** and are therefore correctly present.

A next step for NMEA:
update GxGSA to NMEA 4.10
this will allow to resolve the satellite numbering by introducing the standard character used by the RTCM specifications for the various constellations:
GPS: "G"
GLONASS: "R"
GALILEO: "E"
BEIDOU: "C"
QZSS: "J"
IRNSS/NAVIC: "I"
in addition to the satellite number
(this will avoid double numbering between GPS and GALILEO, currently not resolved in QGis),
it must be said that it will be more readable satellitesInView and satellitesInUse (satprn list) and the Skyplot graphs
Ciao

nyalldawson commented on external/nmea/parse.c:
```diff
@@ -356,31 +356,69 @@ int nmea_parse_GPGSA( const char *buff, int buff_sz, nmeaGPGSA *pack )
+    ptr++;
+  }
+
+  // Check if the number of fields is even, indicating the presence of SIGNAL_ID
+  if ( field_count % 2 == 0 )
+  {
+    has_signal_id = 1;
+  }
```
Why not just:

```suggestion
  has_signal_id = field_count >= 21 ? 1 : 0;
```

nyalldawson commented on src/app/gps/qgsgpsinformationwidget.cpp:
```diff
@@ -367,10 +368,10 @@ void QgsGpsInformationWidget::displayGPSInformation( const QgsGpsInformation &in
-  {
-    mPlot->setAxisScale( QwtPlot::xBottom, 0, info.satellitesInView.size() );
-  } //signal
+  // if ( mStackedWidget->currentIndex() == 1 && info.satInfoComplete ) //signal
+  // {
+  //   mPlot->setAxisScale( QwtPlot::xBottom, 0, info.satellitesInView.size() );
+  // } //signal
```
```suggestion
```

nyalldawson commented on src/app/gps/qgsgpsinformationwidget.cpp:
```diff
@@ -548,7 +588,7 @@ void QgsGpsInformationWidget::displayGPSInformation( const QgsGpsInformation &in
     mTxtQuality->setText( info.qualityDescription() );
     mTxtSatellitesUsed->setText( tr( "%1 used (%2 in view)" ).arg( info.satellitesUsed ).arg( info.satellitesInView.size() ) );
-    mTxtStatus->setText( info.status == 'A' ? tr( "Valid" ) : info.status == 'V' ? tr( "Invalid" ) : QString() );
+    mTxtStatus->setText( info.status == 'A' ? tr( "Valid" ) : info.status == 'V' ? tr( "Invalid" ) : QString( "Not AV (%1)" ).arg( info.status ) );
```
```suggestion
    mTxtStatus->setText( info.status == 'A' ? tr( "Valid" ) : info.status == 'V' ? tr( "Invalid" ) : tr( "Other (%1)" ).arg( info.status ) );
```

nyalldawson commented on src/core/gps/qgsnmeaconnection.cpp:
```diff
@@ -451,11 +453,20 @@ void QgsNmeaConnection::processGsvSentence( const char *data, int len )
       if ( mLastGPSInformation.satellitesInView.size() > NMEA_SATINPACK )
       {
-        for ( const QgsSatelliteInfo &existingSatInView : std::as_const( mLastGPSInformation.satellitesInView ) )
+        for ( QgsSatelliteInfo &existingSatInView : mLastGPSInformation.satellitesInView )
```
I don't think this approach will work -- it's a qt thing due to implicit sharing of QList, but you're not allowed to modify the list directly while iterating over it using a range based loop.

For this to work you'd need to change to:

```suggestion
        for ( int i = 0; i < mLastGPSInformation.satellitesInView.size(); ++i )
        {
          QgsSatelliteInfo &existingSatInView = mLastGPSInformation.satellitesInView[i];
```

bettellam committed:
Update qgsgpsinformation.cpp

Update for Status not A, V  - for UM982 Status = D - Differential when it becomes RTKfix

bettellam commented:
@nyalldawson test Windows Qt6 builds
**Ublox ZED-F9P**
![Screenshot 2024-10-17 122350](https://github.com/user-attachments/assets/c321ccb5-41dd-48aa-951e-145901629c55)

![Screenshot 2024-10-17 122407](https://github.com/user-attachments/assets/cbdc7ff8-60e2-4ee5-8338-abb50e5f613d)

![Screenshot 2024-10-17 122425](https://github.com/user-attachments/assets/5cbdc609-6169-4872-a1cc-0604b4a254f2)

**UNICORE UM982**

![Screenshot 2024-10-17 132724](https://github.com/user-attachments/assets/3ed992f3-b729-4413-a588-eabed6ec5f73)

![Screenshot 2024-10-17 133140](https://github.com/user-attachments/assets/f2fcb904-015a-4806-8d88-3e7832f26e83)


NEW [Update qgsgpsinformation.cpp](https://github.com/qgis/QGIS/pull/58357/commits/873e78f3fb5397e5fcccd88303bc4a11ff147d19)

bettellam commented:
@nyalldawson Help I don't understand what I have to do to get a new executable. Sorry I'm not very familiar with github, old age kills my memory :-)

nyalldawson commented:
@bettellam can i get your thoughts on https://github.com/qgis/QGIS/pull/59124 ?

bettellam committed:
Merge branch 'master' into patch-1

bettellam committed:
Update parse.c update talkerId[1] GSV

bettellam committed:
Update qgsgpsinformation.cpp update isValid

bettellam committed:
Update qgsgpsinformation.cpp upsate isValid identation

bettellam committed:
Update qgsgpsinformation.cpp update isValid identation :-)

bettellam commented:
> @bettellam can i get your thoughts on #59124 ?

very good, test done OK!!

bettellam committed:
Update parse.c

bettellam committed:
Update parse.c

bettellam committed:
Update qgsnmeaconnection.cpp RMC Status Differential for UM98x

bettellam committed:
Update qgsnmeaconnection.cpp

bettellam committed:
Update parse.c

bettellam committed:
Update qgsgpsinformationwidget.cpp

bettellam committed:
Update qgsgpsinformationwidget.cpp

bettellam commented:
With ZED-F9P everything is OK!
With UM982 there is a strange behavior in the two graphs:
Signal Plot and SkyView.
The graphs are seen in bursts for very small intervals of time (tenths of a second).
I don't understand the reason ??????

https://github.com/user-attachments/assets/97d237d6-8be7-4b77-b762-4766fb15b791

bettellam committed:
Update qgsnmeaconnection.cpp -other text strings that do not start with '$'

bettellam commented:
small improvement....

https://github.com/user-attachments/assets/ca96722e-4abb-418e-bf16-84995b9641fe

bettellam commented:
@nyalldawson  Ciao
I think I found the reason for the flickering with UM982.
I tested them simultaneously on the same antenna using an antenna splitter.

Currently the widget is redrawn at every NMEA sentence received with `satInfoComplete=true`.

This is the NMEA sequence for **1 epoch with ZED-F9P**:
GGA (satInfoComplete=true)
**GSA** (satInfoComplete=true and `satellitesInView.clear()`)
**GSV** (satInfoComplete=false)
GST (satInfoComplete=true)
ZDA (satInfoComplete=true)
RMC (satInfoComplete=true)
VTG (satInfoComplete=true).

while the NMEA sequence for **1 epoch with UM982** :
GGA (satInfoComplete=true)
ZDA (satInfoComplete=true)
**GSA** (satInfoComplete=true and `satellitesInView.clear()` )
RMC (satInfoComplete=true)
GST (satInfoComplete=true)
VTG (satInfoComplete=true)
**GSV** (satInfoComplete=false).

1)
In ZED-F9P GSA and GSV with contiguous;
in UM982, after GSA has executed `satellitesInView.clear()` , you have to wait for the next GGA to have data representable in the widget.

2)
The UM982 NMEA GSV sentences are much more numerous, as UM982 has a capacity to receive more frequency signals from satellites (just under double compared to ZED-F9P).

3)
in the reading period of an NMEA epoch:
ZED-F9P satInfoComplete=true 25% , satInfoComplete=false 75%;
UM982 satInfoComplete=true 7% , satInfoComplete=false 93%.

**This explains the flickering!**

A possible improvement could be to move the widget **redrawing ONLY at the completion of an NMEA epoch**, i.e. before proceeding with `parseGGA`, generally coinciding with the various GNSS receivers at "**start of new epoch**".
This eventuality must be tested in a subsequent pull request.

I consider this pull request complete for the moment.
Ciao

DelazJ commented:
@bettellam sorry to jump in this way, but are there screenshots we could pick here to feed the QGIS docs, please? Or you'd like to provide to QGIS docs? [Ours are quite poor](https://docs.qgis.org/3.34/en/docs/user_manual/working_with_gps/live_GPS_tracking.html#gps-information-panel) as we don't own/play with GPS in the docs team. Thanks in advance.

bettellam commented:
Hi @DelazJ , the GPS documentation is actually a bit dated compared to the great update work done by  @nyalldawson . 
This PR is about to be closed, so some new implementations will not be available. 
No problem for me if you want to take images from here. 
I remain available if you need anything else. 
bye

bettellam committed:
Merge branch 'master' into patch-1

bettellam committed:
Update qgsgpsinformationwidget.cpp

bettellam committed:
Update qgsgpsinformationwidget.cpp

bettellam committed:
Update qgsgpsinformationwidget.cpp

bettellam commented:
Sorry, I don't understand what's wrong!!!
Help...
Thanks!

bettellam committed:
Update qgsgpsinformationwidget.cpp

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bettellam commented:
I found the reason for the reported errors. 
[https://clang.llvm.org/docs/ClangFormatStyleOptions.html](https://clang.llvm.org/docs/ClangFormatStyleOptions.html)
is it possible to reopen the pull request? thanks

bettellam commented:
@nyalldawson hello, should I consider this PR permanently closed or is there a possibility that it will be merged? 
thanks . Ciao

flytorisan commented:
@bettellam Thank you for your reply and the information provided. 
I found the latest firmware (Build 17548) on SparkFun's GitHub and applied it to my UM980. 
After that, I conducted a survey using UM980 and QGIS, and confirmed that the issue was resolved. 
Thank you very much. Best regards.

bettellam commented:
@flytorisan Perfect!
Thanks for the information, I hope to find a recent firmware for my UM982!
Good luck!
Ciao
